kernel/src/component.rs

```diff
@@ -0,0 +1,30 @@
+/// A Component extends the functionality of the Tock kernel.
+/// This abstraction is intended to make the kernel boot sequence simpler:
+/// without it, the reset_handler involves lots of driver-specific
+/// initialization. The Component trait encapsulates all of the
+/// initialziation and configuration of a kernel extension inside
+/// the finalize function call.
+///
+/// Note that instantiating a component does not necessarily instantiate
+/// the underlying Output type; this can be instantiated when finalize()
+/// is called.
+
+pub trait Component {
+    type Output;
+    unsafe fn finalize(&mut self) -> Self::Output;
```

Comment:
Can `finalize` fail? For example, if a dependency was not provided?

Reply:
I think finalize() can fail, but in that case it should panic.